File: HandOnML/P08_DeepLearning/01_ArtificialNeuralNetwork/ann.py
# Artificial Neural Network

# Installing Theano
# pip install --upgrade --no-deps git+git://github.com/Theano/Theano.git

# Installing Tensorflow
# Install Tensorflow from the website: https://www.tensorflow.org/versions/r0.12/get_started/os_setup.html

# Installing Keras
# pip install --upgrade keras

# Part 1 - Data Preprocessing

# Importing the libraries
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

# Importing the dataset
dataset = pd.read_csv('Churn_Modelling.csv')
X = dataset.iloc[:, 3:13].values
y = dataset.iloc[:, 13].values

# Encoding categorical data
from sklearn.preprocessing import LabelEncoder, OneHotEncoder
labelencoder_X_Country = LabelEncoder()
X[:, 1] = labelencoder_X_Country.fit_transform(X[:, 1])
labelencoder_X_Gender = LabelEncoder()
X[:, 2] = labelencoder_X_Gender.fit_transform(X[:, 2])
onehotencoder = OneHotEncoder(categorical_features = [1])
X = onehotencoder.fit_transform(X).toarray()
X = X[:, 1:] # Avoid dummy variable trap

# Splitting the dataset into the Training set and Test set
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2, random_state = 0)

# Feature Scaling
from sklearn.preprocessing import StandardScaler
sc_X = StandardScaler()
X_train = sc_X.fit_transform(X_train)
X_test = sc_X.transform(X_test)

# Part 2 - Let's make the ANN
# Import the keras library
import keras
from keras.models import Sequential
from keras.layers import Dense

# Initializing the ANN
classifier = Sequential()

# Adding the input layer and first hidden layer
classifier.add(Dense(output_dim = 6, # Tips: choose node numbers of hidden layers equal average of node in both intput and output layers
                                     # Input layers: 11 nodes - 11 independent variables
                                     # Output layers: 1 node - 1 dependent variable
                     init = 'uniform',
                     activation = 'relu', # rectifier activation function for hidden layer
                     input_dim = 11))

# Adding the second hidden layer
classifier.add(Dense(output_dim = 6, # Tips: choose node numbers of hidden layers equal average of node in both intput and output layers
                                     # Input layers: 11 nodes - 11 independent variables
                                     # Output layers: 1 node - 1 dependent variable
                     init = 'uniform',
                     activation = 'relu')) # rectifier activation function for hidden layer

# Adding the output layer
classifier.add(Dense(output_dim = 1, # We have only 1 dependent variable
                     init = 'uniform',
                     activation = 'sigmoid')) # Use sigmoid function for output layer

# Copy the ANN
classifier.compile(optimizer = 'adam', # Optimizer is algorithm used to find initialized set-up weights
                                       # 'adam' is one of the effective stochastic gradient descend
                   loss = 'binary_crossentropy', # loss function equivalent with sigmoid activation function
                   metrics=['accuracy']
                   )

# Fitting the ANN to the Training set
classifier.fit(X_train, y_train, batch_size = 10, nb_epoch = 100)

# Predict the Test set result
y_pred = classifier.predict(X_test)
y_pred = (y_pred > 0.5)

# Making the confusion matrix
from sklearn.metrics import confusion_matrix
cm = confusion_matrix(y_test, y_pred)

File: HandOnML/P08_DeepLearning/01_ArtificialNeuralNetwork/ann.R
# Artificial Neural Network

# Importing the dataset
dataset = read.csv('Churn_Modelling.csv')
dataset = dataset[,4:14]

# Encoding the categorical variables as factors
# For some deep learning package, it only working with numerical variables -> Convert 
#   categorical variable to numeric variables
dataset$Geography = as.numeric(factor(dataset$Geography,
                              levels = c('France', 'Spain', 'Germany'),
                              labels = c(1, 2, 3)))
dataset$Gender = as.numeric(factor(dataset$Gender,
                              levels = c('Female', 'Male'),
                              labels = c(1, 2)))

# Splitting the dataset into the Training set and Test set
# install.packages('caTools')
library(caTools)
set.seed(123)
split = sample.split(dataset$Exited, SplitRatio = 0.8)
training_set = subset(dataset, split == TRUE)
test_set = subset(dataset, split == FALSE)

# Feature Scaling
# For colum that categorical trasnformed by factor function, these are not numeric so can't be scale
training_set[-11] = scale(training_set[-11])
test_set[-11] = scale(test_set[-11])

# Fitting ANN to the Traning set
# install.packages('h2o')
library(h2o)
h2o.init(nthreads = -1)
classifier = h2o.deeplearning(y = 'Exited',
                              training_frame = as.h2o(training_set), # convert dataframe to h2o training frame
                              activation = 'Rectifier',
                              hidden = c(2,6), # 2 hidden layers and each layer has 6 nodes
                              epochs = 100,
                              train_samples_per_iteration = -2) # batch size will be auto tunned

# Predicting the Test set results
prob_pred = h2o.predict(classifier, newdata = as.h2o(test_set[-11]))
y_pred = (prob_pred > 0.5)
y_pred = as.vector(y_pred)

# Making the confusion matrix
cm = table(test_set[,11], y_pred)

# Disconnnecting from h2o servers
h2o.shutdown()